File: reports/20260826-143732/results/identity/data-plane/python/auth/identity-dp-python-credential-chain/python-azure-tools/with-azure-tools/generated-code/environment_detector.py
"""Detect the deployment environment from well-known environment variables."""

from __future__ import annotations

import os
from dataclasses import dataclass
from enum import Enum
from typing import Mapping


class DeploymentEnvironment(str, Enum):
    DEV = "dev"
    CI = "ci"
    PRODUCTION = "production"


@dataclass(frozen=True)
class DetectionResult:
    environment: DeploymentEnvironment
    reason: str


_CI_MARKERS = {
    "TF_BUILD": "Azure Pipelines",
    "GITHUB_ACTIONS": "GitHub Actions",
    "GITLAB_CI": "GitLab CI",
    "JENKINS_URL": "Jenkins",
    "TEAMCITY_VERSION": "TeamCity",
    "BITBUCKET_BUILD_NUMBER": "Bitbucket Pipelines",
    "CI_PROJECT_DIR": "CI project workspace",
    "PIPELINE_WORKSPACE": "pipeline workspace",
}

_PRODUCTION_MARKERS = {
    "IDENTITY_ENDPOINT": "Azure managed identity endpoint",
    "MSI_ENDPOINT": "Azure managed identity endpoint",
    "IMDS_ENDPOINT": "Azure Instance Metadata Service endpoint",
    "WEBSITE_INSTANCE_ID": "Azure App Service host",
    "FUNCTIONS_WORKER_RUNTIME": "Azure Functions host",
    "CONTAINER_APP_NAME": "Azure Container Apps host",
}


def detect_environment(
    environ: Mapping[str, str] | None = None,
) -> DetectionResult:
    """Classify the current process as development, CI, or production."""
    values = os.environ if environ is None else environ

    override = values.get("APP_ENVIRONMENT", "").strip().lower()
    if override:
        aliases = {
            "dev": DeploymentEnvironment.DEV,
            "development": DeploymentEnvironment.DEV,
            "ci": DeploymentEnvironment.CI,
            "pipeline": DeploymentEnvironment.CI,
            "prod": DeploymentEnvironment.PRODUCTION,
            "production": DeploymentEnvironment.PRODUCTION,
        }
        if override not in aliases:
            expected = ", ".join(sorted(aliases))
            raise ValueError(
                f"Invalid APP_ENVIRONMENT value {override!r}; expected one of: {expected}"
            )
        return DetectionResult(aliases[override], "APP_ENVIRONMENT override")

    for variable, description in _CI_MARKERS.items():
        if values.get(variable):
            return DetectionResult(
                DeploymentEnvironment.CI,
                f"{description} detected from {variable}",
            )

    workload_identity_variables = (
        "AZURE_TENANT_ID",
        "AZURE_CLIENT_ID",
        "AZURE_FEDERATED_TOKEN_FILE",
    )
    if all(values.get(variable) for variable in workload_identity_variables):
        return DetectionResult(
            DeploymentEnvironment.PRODUCTION,
            "Kubernetes workload identity configuration detected",
        )

    for variable, description in _PRODUCTION_MARKERS.items():
        if values.get(variable):
            return DetectionResult(
                DeploymentEnvironment.PRODUCTION,
                f"{description} detected from {variable}",
            )

    return DetectionResult(
        DeploymentEnvironment.DEV,
        "no CI or Azure-hosted production markers were found",
    )

File: rerun-reports/03-java/20260829-075759/results/identity/data-plane/java/auth/identity-dp-java-credential-chain/java-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/main/java/com/example/azureauth/EnvironmentDetector.java
package com.example.azureauth;

import java.util.Locale;
import java.util.Map;
import java.util.Set;

public final class EnvironmentDetector {
    private static final Set<String> CI_MARKERS = Set.of(
        "CI",
        "TF_BUILD",
        "GITHUB_ACTIONS",
        "GITLAB_CI",
        "JENKINS_URL",
        "BUILD_SOURCESDIRECTORY",
        "PIPELINE_WORKSPACE"
    );

    private static final Set<String> PRODUCTION_MARKERS = Set.of(
        "IDENTITY_ENDPOINT",
        "MSI_ENDPOINT",
        "IMDS_ENDPOINT",
        "WEBSITE_INSTANCE_ID",
        "CONTAINER_APP_NAME",
        "AZURE_FEDERATED_TOKEN_FILE"
    );

    private final Map<String, String> environment;

    public EnvironmentDetector() {
        this(System.getenv());
    }

    EnvironmentDetector(Map<String, String> environment) {
        this.environment = Map.copyOf(environment);
    }

    public DeploymentEnvironment detect() {
        String override = environment.get("APP_ENVIRONMENT");
        if (hasText(override)) {
            return parseOverride(override);
        }
        if (hasAnyMarker(CI_MARKERS)) {
            return DeploymentEnvironment.CI;
        }
        if (hasAnyMarker(PRODUCTION_MARKERS) || hasWorkloadIdentityConfiguration()) {
            return DeploymentEnvironment.PRODUCTION;
        }
        return DeploymentEnvironment.DEVELOPMENT;
    }

    private boolean hasAnyMarker(Set<String> names) {
        return names.stream().map(environment::get).anyMatch(EnvironmentDetector::hasText);
    }

    private boolean hasWorkloadIdentityConfiguration() {
        return hasText(environment.get("AZURE_TENANT_ID"))
            && hasText(environment.get("AZURE_CLIENT_ID"))
            && hasText(environment.get("AZURE_FEDERATED_TOKEN_FILE"));
    }

    private static DeploymentEnvironment parseOverride(String value) {
        return switch (value.trim().toLowerCase(Locale.ROOT)) {
            case "dev", "development", "local" -> DeploymentEnvironment.DEVELOPMENT;
            case "ci", "pipeline" -> DeploymentEnvironment.CI;
            case "prod", "production" -> DeploymentEnvironment.PRODUCTION;
            default -> throw new IllegalArgumentException(
                "APP_ENVIRONMENT must be one of: dev, ci, or production"
            );
        };
    }

    private static boolean hasText(String value) {
        return value != null && !value.isBlank();
    }
}
